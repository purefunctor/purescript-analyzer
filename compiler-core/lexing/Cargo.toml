[package]
name = "lexing"
version = "0.1.0"
edition = "2021"

[dependencies]
memchr = "2.7.4"
syntax = { version = "0.1.0", path = "../syntax" }
unicode_categories = "0.1.1"

[dev-dependencies]
insta = "1.41.1"
test_each_file = "0.3.4"

