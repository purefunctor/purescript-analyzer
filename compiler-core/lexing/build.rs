fn main() {
    println!("cargo:rerun-if-changed=tests/layout");
    println!("cargo:rerun-if-changed=tests/lexer");
}
