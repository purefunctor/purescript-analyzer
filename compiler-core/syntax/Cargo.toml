[package]
name = "syntax"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
hashbrown = "0.15.2"
la-arena = "0.3.1"
paste = "1.0.15"
rowan = "0.16.1"
rustc-hash = "2.1.1"
