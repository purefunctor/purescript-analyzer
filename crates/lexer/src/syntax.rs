#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
#[allow(non_camel_case_types)]
#[repr(u16)]
pub enum SyntaxKind {
    // region: whitespace
    WHITESPACE,
    LINE_COMMENT,
    BLOCK_COMMENT,
    // endregion: whitespace

    // region: keywords
    /// _
    UNDERSCORE,
    /// ado
    ADO,
    /// as
    AS,
    /// case
    CASE,
    /// class
    CLASS,
    /// data
    DATA,
    /// derive
    DERIVE,
    /// do
    DO,
    /// else
    ELSE,
    /// false
    FALSE,
    /// foreign
    FOREIGN,
    /// hiding
    HIDING,
    /// if
    IF,
    /// import
    IMPORT,
    /// in
    IN,
    /// infix
    INFIX,
    /// infixl
    INFIXL,
    /// infixr
    INFIXR,
    /// instance
    INSTANCE,
    /// let
    LET,
    /// module
    MODULE,
    /// newtype
    NEWTYPE,
    /// of
    OF,
    /// then
    THEN,
    /// true
    TRUE,
    /// type
    TYPE,
    /// where
    WHERE,
    // endregion: keywords

    // region: names
    /// Module names
    MODULE_NAME,
    /// Named holes
    SOURCE_HOLE,
    /// Lower-case name
    SOURCE_IDENTIFIER,
    /// Upper-case name
    SOURCE_PROPER,
    /// +, -, *, /
    SOURCE_OPERATOR,
    /// (+), (-), (*), (/)
    SOURCE_SYMBOL,
    // endregion: names

    // region: operators
    /// <-, ←
    LEFT_ARROW,
    /// <=, ⇐
    LEFT_DOUBLE_ARROW,
    /// ->, →
    RIGHT_ARROW,
    /// =>, ⇒
    RIGHT_DOUBLE_ARROW,
    /// ::, ∷
    DOUBLE_COLON,
    /// forall, ∀
    FORALL,
    /// =
    EQUALS,
    /// .
    DOT,
    /// ,
    COMMA,
    /// \\
    BACKSLASH,
    /// |
    PIPE,
    /// @
    AT,
    /// `
    TICK,
    // endregion: operators

    // region: delimiters
    /// (
    LEFT_PARENTHESIS,
    /// )
    RIGHT_PARENTHESIS,
    /// [
    LEFT_SQUARE,
    /// ]
    RIGHT_SQUARE,
    /// {
    LEFT_CURLY,
    /// }
    RIGHT_CURLY,
    // endregion: delimiters

    // region: values
    /// ""
    STRING,
    /// """
    RAW_STRING,
    /// 'a', '\n', '\xFFFFFF'
    CHAR,
    /// 0, 0xFF
    INTEGER,
    /// 0.0, 1.23e+5
    NUMBER,
    // endregion: values

    // region: control
    END_OF_FILE,
    LAYOUT_START,
    LAYOUT_DELIMITER,
    LAYOUT_END,
    // endregion: control
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct PureScript {}

impl rowan::Language for PureScript {
    type Kind = SyntaxKind;

    fn kind_from_raw(raw: rowan::SyntaxKind) -> Self::Kind {
        assert!(raw.0 <= SyntaxKind::END_OF_FILE as u16);
        unsafe { std::mem::transmute::<u16, SyntaxKind>(raw.0) }
    }

    fn kind_to_raw(kind: Self::Kind) -> rowan::SyntaxKind {
        rowan::SyntaxKind(kind as u16)
    }
}

impl From<SyntaxKind> for rowan::SyntaxKind {
    fn from(value: SyntaxKind) -> Self {
        Self(value as u16)
    }
}

pub type SyntaxNode = rowan::SyntaxNode<PureScript>;
pub type SyntaxNodeChildren = rowan::SyntaxNodeChildren<PureScript>;
pub type SyntaxNodePtr = rowan::ast::SyntaxNodePtr<PureScript>;
pub type SyntaxToken = rowan::SyntaxToken<PureScript>;
pub type SyntaxElement = rowan::SyntaxElement<PureScript>;

impl SyntaxKind {
    pub fn is_whitespace_or_comment(self) -> bool {
        matches!(
            self,
            SyntaxKind::WHITESPACE | SyntaxKind::LINE_COMMENT | SyntaxKind::BLOCK_COMMENT
        )
    }
}
