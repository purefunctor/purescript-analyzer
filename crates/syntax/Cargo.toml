[package]
name = "syntax"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
rowan = "0.15.15"
smol_str = "0.2.0"
