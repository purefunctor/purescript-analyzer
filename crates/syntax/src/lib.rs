pub mod ast;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
#[repr(u16)]
pub enum SyntaxKind {
    Whitespace = 0,
    LineComment,
    BlockComment,

    Module,
    ModuleHeader,
    ModuleImports,
    ModuleBody,
    ModuleKw,
    WhereKw,

    ExportList,
    ExportTypeOp,
    ExportClass,
    ExportModule,
    ExportOp,
    ExportValue,
    ExportType,

    ImportList,
    ImportOp,
    ImportType,
    DataAll,
    DataEnumerated,
    ImportTypeOp,
    ImportClass,
    ImportValue,

    ImportDeclaration,
    ImportKw,
    ImportHidden,
    HidingKw,
    ImportQualified,
    AsKw,

    ModuleName,
    QualifiedName,
    QualifiedPrefix,
    Name,
    NameRef,
    Upper,
    Lower,
    Label,
    Hole,
    Operator,

    At,
    Backslash,
    Colon,
    Colon2,
    Comma,
    Equal,
    Minus,
    Period,
    Period2,
    Pipe,
    Question,
    Tick,
    Underscore,

    LeftArrow,
    RightArrow,
    LeftThickArrow,
    RightThickArrow,
    LeftParenthesis,
    RightParenthesis,
    LeftCurly,
    RightCurly,
    LeftSquare,
    RightSquare,

    AdoExpression,
    QualifiedAdo,
    AdoKw,

    DoExpression,
    QualifiedDo,
    // UnqualifiedDo,
    DoKw,

    DoBind,
    DoLetBinding,
    DoDiscard,

    ApplicationExpression,
    TermArgument,
    TypeArgument,

    IfThenElseExpression,
    IfKw,
    ThenKw,
    ElseKw,

    LetInExpression,
    LetKw,
    InKw,
    LetBindingSignature,
    LetBindingName,
    LetBindingPattern,
    WhereExpression,

    CaseKw,
    OfKw,
    CaseBranch,
    CaseExpression,

    ExpressionInfixChain,
    ExpressionOperatorChain,
    NegateExpression,

    ConstructorExpression,
    OperatorNameExpression,
    VariableExpression,

    SectionExpression,
    LambdaExpression,
    LiteralExpression,
    ParenthesizedExpression,
    RecordAccessExpression,
    RecordUpdateExpression,
    RecordUpdateLeaf,
    RecordUpdateBranch,
    TypedExpression,

    LiteralChar,
    LiteralString,
    LiteralRawString,
    LiteralInteger,
    LiteralNumber,
    LiteralTrue,
    LiteralFalse,
    LiteralArray,
    LiteralRecord,
    RecordPun,
    RecordField,

    ForallType,
    TypeVariableKinded,
    TypeVariableName,

    ArrowType,
    ConstrainedType,

    ApplicationType,
    ConstructorType,
    IntegerType,
    KindedType,
    OperatorNameType,
    ParenthesizedType,
    RecordType,
    RowType,
    RowInner,
    RowField,
    RowTail,
    StringType,
    TypeOperatorChain,
    VariableType,
    WildcardType,

    BinderOperatorChain,
    ConstructorBinder,
    LiteralBinder,
    NegativeBinder,
    ParenthesizedBinder,
    TypedBinder,
    VariableBinder,
    WildcardBinder,

    ValueEquationDeclaration,
    ValueAnnotationDeclaration,

    DataAnnotation,
    DataDeclaration,
    DataVariables,
    DataConstructors,
    DataConstructor,
    ConstructorFields,
    DataKw,

    NewtypeDeclarationAnnotation,
    NewtypeDeclaration,
    NewtypeKw,
    ForallKw,

    TypeDeclarationAnnotation,
    TypeDeclarationRole,
    TypeDeclaration,
    TypeKw,
    RoleKw,
    NominalKw,
    RepresentationalKw,
    PhantomKw,

    ClassSignature,
    ClassDeclaration,
    ClassConstraints,
    ClassVariables,
    ClassFundeps,
    FundepDetermined,
    FundepDetermines,
    FundepVariables,
    ClassMembers,
    ClassMember,
    ClassKw,

    InstanceChain,
    InstanceDeclaration,
    InstanceAssertions,
    InstanceMemberSignature,
    InstanceMemberEquation,
    InstanceKw,

    DeriveNewtypeDeclaration,
    DeriveInstanceDeclaration,
    DeriveKw,

    ForeignDataDeclaration,
    ForeignValueDeclaration,
    ForeignKw,

    FixityDeclaration,
    InfixlKw,
    InfixrKw,
    InfixKw,

    UnconditionalBinding,
    GuardedBinding,
    PatternGuard,
    GuardedExpression,

    /// Convenience node for patterns such as `l: e` or `e :: T`.
    Labeled,
    /// Convenience node for patterns such as `@variable` or `?hole`.
    Prefixed,
    /// Convenience node for patterns such as `( element )`.
    Wrapped,
    /// Convenience node for a non-empty array of elements.
    OneOrMore,
    /// Convenience node for an array of elements.
    ZeroOrMore,
    /// Convenience node for a tuple of elements.
    Pair,
    /// Convenience node for patterns such as `a , b , c`
    Separated,

    LayoutStart,
    LayoutSep,
    LayoutEnd,

    Sentinel,
    Source,
    Error,
    EndOfFile,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct PureScript {}

impl rowan::Language for PureScript {
    type Kind = SyntaxKind;

    fn kind_from_raw(raw: rowan::SyntaxKind) -> Self::Kind {
        assert!(raw.0 <= SyntaxKind::EndOfFile as u16);
        unsafe { std::mem::transmute::<u16, SyntaxKind>(raw.0) }
    }

    fn kind_to_raw(kind: Self::Kind) -> rowan::SyntaxKind {
        rowan::SyntaxKind(kind as u16)
    }
}

impl From<SyntaxKind> for rowan::SyntaxKind {
    fn from(value: SyntaxKind) -> Self {
        Self(value as u16)
    }
}

pub type SyntaxNode = rowan::SyntaxNode<PureScript>;
pub type SyntaxNodeChildren = rowan::SyntaxNodeChildren<PureScript>;
pub type SyntaxNodePtr = rowan::ast::SyntaxNodePtr<PureScript>;
pub type SyntaxToken = rowan::SyntaxToken<PureScript>;
pub type SyntaxElement = rowan::SyntaxElement<PureScript>;

impl SyntaxKind {
    pub fn is_whitespace_or_comment(&self) -> bool {
        matches!(self, Self::Whitespace | Self::LineComment | Self::BlockComment)
    }

    pub fn is_operator(&self) -> bool {
        matches!(self, Self::Operator | Self::Minus | Self::Colon | Self::LeftThickArrow | Self::RightThickArrow)
    }

    pub fn is_reserved_operator(&self) -> bool {
        matches!(self, Self::RightArrow | Self::RightThickArrow)
    }

    pub fn is_lower(&self) -> bool {
        matches!(self, Self::Lower | Self::AsKw)
    }

    pub fn is_label(&self) -> bool {
        matches!(
            self,
            SyntaxKind::ModuleKw
                | SyntaxKind::WhereKw
                | SyntaxKind::ImportKw
                | SyntaxKind::AsKw
                | SyntaxKind::AdoKw
                | SyntaxKind::DoKw
                | SyntaxKind::IfKw
                | SyntaxKind::ThenKw
                | SyntaxKind::ElseKw
                | SyntaxKind::LetKw
                | SyntaxKind::InKw
                | SyntaxKind::CaseKw
                | SyntaxKind::OfKw
                | SyntaxKind::DataKw
                | SyntaxKind::NewtypeKw
                | SyntaxKind::ForallKw
                | SyntaxKind::TypeKw
                | SyntaxKind::ClassKw
                | SyntaxKind::InstanceKw
                | SyntaxKind::DeriveKw
                | SyntaxKind::ForeignKw
                | SyntaxKind::InfixlKw
                | SyntaxKind::InfixrKw
                | SyntaxKind::InfixKw
                | SyntaxKind::LiteralString
                | SyntaxKind::LiteralRawString
                | SyntaxKind::Lower
        )
    }

    pub fn is_end(&self) -> bool {
        matches!(self, Self::LayoutSep | Self::LayoutEnd | Self::EndOfFile)
    }

    pub fn is_layout(&self) -> bool {
        matches!(self, Self::LayoutStart | Self::LayoutSep | Self::LayoutEnd)
    }
}
