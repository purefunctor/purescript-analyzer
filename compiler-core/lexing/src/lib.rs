mod categories;
mod layout;
mod lexed;
mod lexer;

pub use lexed::Lexed;
use syntax::SyntaxKind;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Position {
    pub line: u32,
    pub column: u32,
}

pub fn lex(source: &str) -> Lexed<'_> {
    let mut lexer = lexer::Lexer::new(source);
    while !lexer.is_eof() {
        lexer.take_token();
    }
    lexer.finish()
}

pub fn layout(lexed: &Lexed) -> Vec<SyntaxKind> {
    let mut layout = layout::Layout::new(lexed);
    while !layout.is_eof() {
        layout.take_token();
    }
    layout.finish()
}
