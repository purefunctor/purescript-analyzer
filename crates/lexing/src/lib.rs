//! A lexer for PureScript source code.

mod layout;
mod lexed;
mod lexer;

pub use lexed::Lexed;

use layout::Machine;
use lexer::Lexer;
use syntax::SyntaxKind;

/// Tokenizes a source string.
pub fn lex(source: &str) -> Lexed {
    let mut lexer = Lexer::new(source);
    loop {
        if lexer.is_eof() {
            break lexer.finalize();
        }
        lexer.take_token();
    }
}

/// Applies the layout algorithm.
pub fn layout(lexed: &Lexed) -> Vec<SyntaxKind> {
    let mut machine = Machine::new(lexed);
    loop {
        if machine.is_eof() {
            break machine.finalize();
        }
        machine.take_token();
    }
}

#[cfg(test)]
mod tests {
    use syntax::SyntaxKind;

    use crate::{layout, lex};

    const FULL_SOURCE: &str = r"module Test where

test = ado
    baz
    let foo = bar
    in bar

test = ado in foo

test = ado
    foo <- bar $ let a = 42 in a
    baz <- b
    in bar

test = ado
    foo
    let bar = let a = 42 in a
    let baz = 42
    in bar
    
example1 = do
  foo bar
  <|> baz

example2 = do
  foo bar
  `wat` baz

example3 =
  case _ of
    Foo a -> 1
    Bar b -> 2
    `append` 3

example4 =
  case _ of
    Foo a -> 1
    Bar b -> 2
    + 3

-- Including data because of `|` masking
data Foo
 = Foo
 | Bar
 | Baz

test =
  case foo of
    a | b, c ->
      d
    a | b, c -> d

test = case a, b of
  c, d
   | e ->
     case e of
       f | true -> bar
         | false -> baz
   | f -> g

test a
  | false =
      case false of
        true | a > 12 -> true
  | otherwise = true

test = case a of foo | foo \a -> a -> true

test = a `case _ of x | unit # \_ -> true, true -> const` b

test = case a of
  12 | do that
          that   -> this
     | otherwise -> this

test a b = [ case _ of
  12 | case a, b of
         _, 42 -> b
         _, 12 -> false, b -> true
     | case a, b of
         _, 42 -> b
         _, 12 -> false, b -> true, false ]

test a
  | case a, b of
         _, 42 -> b
         _, 12 -> false, b = true
  | case a, b of
         _, 42 -> b
         _, 12 -> false, b = true

test = case foo of
  Nothing -> a
    where a = 12
  Just a -> do
    what
  where
    foo = bar

test = case f of Foo -> do that
                        where foo = 12

import Foo (class Foo)

class Foo a b c d | a -> b, c -> d where
  foo :: Foo

class Foo a b c d | a -> b, c -> d

instance foo :: Foo

test =
  [ case do foo, bar of
      a | b, c -> d, bar
    ]

test =
  [ case do foo, bar of a | b, c -> d, bar ]

test =
  [ do do do foo, bar ]

test =
  [ \foo -> foo, bar ]

test = foo where
  bar =
    case a, b of
      c, d | d == [case true, w of 1, a -> true, false ] -> d
      e, d | do what, do that -> d

test1 = a
test2 = {
  b
}
test3 = do
  foo
  bar (
    baz
  ) == 12
  baz
test4 = c

test = do
  let foo = bar
  foo

test = do
  let foo = bar
  in baz
  foo

test = do
  let foo = bar
    in baz
  foo

test = do
  foo
  foo do
    bar
  <|> bar

test =
  do
    do do
        foo where bar = baz

foo = do
  if true then
    false
  else if false then do
    that
    else do
     what
  that

instance foo :: Foo Int else bar :: Foo String
else baz :: Foo Boolean

type IntType = (-1)

type IntType' = (-
  -- here's a comment
  1)

test =
  let
    foo
      | bar
      , baz =
        42
      | otherwise = 100
  in
    foo

test = do
  let
    foo
      | bar
      , baz =
        42
      | otherwise = 100
  foo

test = ado
  let
    foo
      | bar
      , baz =
        42
      | otherwise = 100
  foo";

    #[test]
    fn tokens_are_identical() {
        let lexed = lex(FULL_SOURCE);

        let mut left = vec![];
        for index in 0..lexed.len() {
            let kind = lexed.kind(index);
            if kind.is_whitespace_or_comment() {
                continue;
            }
            left.push(kind);
        }

        let right: Vec<_> = layout(&lexed)
            .into_iter()
            .filter(|kind| {
                !matches!(
                    kind,
                    SyntaxKind::LayoutStart | SyntaxKind::LayoutSep | SyntaxKind::LayoutEnd
                )
            })
            .collect();

        assert_eq!(left, right);
    }
}
