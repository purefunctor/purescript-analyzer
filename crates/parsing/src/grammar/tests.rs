use std::{fmt::Write, fs, path::PathBuf};

use lexing::{layout, lex};
use syntax::SyntaxKind;

use crate::{
    grammar::{rules::expr_0, rules::module, rules::pat_0, rules::type_0},
    parser::{Event, Parser},
};

fn expect_parse<F, T>(path: PathBuf, rule: F)
where
    F: Fn(&mut Parser) -> T,
{
    let source = fs::read_to_string(&path).unwrap();
    let name = path.file_stem().unwrap().to_str().unwrap();
    let lexed = lex(&source);
    let input = layout(&lexed);
    let mut parser = Parser::new(&input);
    let _ = rule(&mut parser);

    let mut result = String::new();
    writeln!(result, "Input: {source}").unwrap();

    let mut indentation = 0;
    for actual in parser.finalize() {
        if let Event::Start { kind: SyntaxKind::Sentinel } = actual {
            continue;
        }
        match actual {
            Event::Start { .. } => {
                writeln!(result, "{:indentation$}{:?}", "", actual, indentation = indentation)
                    .unwrap();
                indentation += 2
            }
            Event::Finish => {
                indentation -= 2;
                writeln!(result, "{:indentation$}{:?}", "", actual, indentation = indentation)
                    .unwrap();
            }
            _ => {
                writeln!(result, "{:indentation$}{:?}", "", actual, indentation = indentation)
                    .unwrap();
            }
        }
    }

    insta::assert_snapshot!(name, result);
}

#[test_each::path(glob = "crates/parsing/src/grammar/inputs/passing/expression/*.input")]
fn test_expression(path: PathBuf) {
    expect_parse(path, expr_0);
}

#[test_each::path(glob = "crates/parsing/src/grammar/inputs/failing/expression/*.input")]
fn test_expression_failing(path: PathBuf) {
    expect_parse(path, expr_0);
}

#[test_each::path(glob = "crates/parsing/src/grammar/inputs/passing/type/*.input")]
fn test_type(path: PathBuf) {
    expect_parse(path, type_0);
}

#[test_each::path(glob = "crates/parsing/src/grammar/inputs/passing/pattern/*.input")]
fn test_pattern(path: PathBuf) {
    expect_parse(path, pat_0);
}

#[test_each::path(glob = "crates/parsing/src/grammar/inputs/passing/file/*.input")]
fn test_file(path: PathBuf) {
    expect_parse(path, module);
}
