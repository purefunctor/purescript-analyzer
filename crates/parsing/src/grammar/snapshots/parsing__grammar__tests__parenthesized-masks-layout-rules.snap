---
source: crates/parsing/src/grammar/tests.rs
expression: result
---
Input: do
  (
this
should
not
be
legal
but
it
is
  )
  [
this,
as,
well
  ]

Start { kind: DoExpression }
  Start { kind: QualifiedDo }
    Token { kind: DoKw }
  Finish
  Token { kind: LayoutStart }
  Start { kind: LayoutList }
    Start { kind: DoDiscard }
      Start { kind: ParenthesizedExpression }
        Token { kind: LeftParenthesis }
        Start { kind: ApplicationExpression }
          Start { kind: VariableExpression }
            Start { kind: QualifiedName }
              Start { kind: NameRef }
                Token { kind: Lower }
              Finish
            Finish
          Finish
          Start { kind: ArgumentList }
            Start { kind: TermArgument }
              Start { kind: VariableExpression }
                Start { kind: QualifiedName }
                  Start { kind: NameRef }
                    Token { kind: Lower }
                  Finish
                Finish
              Finish
            Finish
            Start { kind: TermArgument }
              Start { kind: VariableExpression }
                Start { kind: QualifiedName }
                  Start { kind: NameRef }
                    Token { kind: Lower }
                  Finish
                Finish
              Finish
            Finish
            Start { kind: TermArgument }
              Start { kind: VariableExpression }
                Start { kind: QualifiedName }
                  Start { kind: NameRef }
                    Token { kind: Lower }
                  Finish
                Finish
              Finish
            Finish
            Start { kind: TermArgument }
              Start { kind: VariableExpression }
                Start { kind: QualifiedName }
                  Start { kind: NameRef }
                    Token { kind: Lower }
                  Finish
                Finish
              Finish
            Finish
            Start { kind: TermArgument }
              Start { kind: VariableExpression }
                Start { kind: QualifiedName }
                  Start { kind: NameRef }
                    Token { kind: Lower }
                  Finish
                Finish
              Finish
            Finish
            Start { kind: TermArgument }
              Start { kind: VariableExpression }
                Start { kind: QualifiedName }
                  Start { kind: NameRef }
                    Token { kind: Lower }
                  Finish
                Finish
              Finish
            Finish
            Start { kind: TermArgument }
              Start { kind: VariableExpression }
                Start { kind: QualifiedName }
                  Start { kind: NameRef }
                    Token { kind: Lower }
                  Finish
                Finish
              Finish
            Finish
          Finish
        Finish
        Token { kind: RightParenthesis }
      Finish
    Finish
    Token { kind: LayoutSep }
    Start { kind: DoDiscard }
      Start { kind: LiteralExpression }
        Start { kind: LiteralArray }
          Token { kind: LeftSquare }
          Start { kind: VariableExpression }
            Start { kind: QualifiedName }
              Start { kind: NameRef }
                Token { kind: Lower }
              Finish
            Finish
          Finish
          Token { kind: Comma }
          Start { kind: VariableExpression }
            Start { kind: QualifiedName }
              Start { kind: NameRef }
                Token { kind: Lower }
              Finish
            Finish
          Finish
          Token { kind: Comma }
          Start { kind: VariableExpression }
            Start { kind: QualifiedName }
              Start { kind: NameRef }
                Token { kind: Lower }
              Finish
            Finish
          Finish
          Token { kind: RightSquare }
        Finish
      Finish
    Finish
  Finish
  Token { kind: LayoutEnd }
Finish
